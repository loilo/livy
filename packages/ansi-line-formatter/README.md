# `@livy/ansi-line-formatter`

This [Livy](../../README.md#readme) formatter formats log records as single lines with terminal highlighting.

## Basic Example

```js
import { FileHandler } from '@livy/file-handler'
import { AnsiLineFormatter } from '@livy/ansi-line-formatter'

const handler = new FileHandler('logs.txt', {
  formatter: new AnsiLineFormatter()
})
```

## Installation

Install it via npm:

```bash
npm install @livy/ansi-line-formatter
```

## Options

An object of options can be passed to this formatter's constructor.

The following options are available:

### `decorated`

**Type:** `boolean | undefined`

**Default:** `undefined`

**Description:** Whether records should be highlighted with ANSI escape sequences. If this option is not provided (or explicitly set to `undefined`), color support will be detected automatically.

### `ignoreEmptyContext`

**Type:** `boolean`

**Default:** `false`

**Description:** Whether to omit empty context objects (only if extra is empty as well).

### `ignoreEmptyExtra`

**Type:** `boolean`

**Default:** `true`

**Description:** Whether to omit empty `extra` objects.

### `include`

**Type:** [`Partial<IncludedRecordProperties>`](../util/README.md#includedrecordproperties)

**Default:** `{}`

**Description:** Which log record properties to include in the output. The passed object is merged into the following default:

```js
{
  datetime: true,
  channel: false,
  level: true,
  severity: false,
  message: true,
  context: true,
  extra: true
}
```

## Public API

### `decorated`

Whether records should be highlighted with ANSI escape sequences. Initially set through the `decorated` option.

### `ignoreEmptyContext`

Whether to omit empty context objects (only if extra is empty as well). Initially set through the `ignoreEmptyContext` option.

### `ignoreEmptyExtra`

Whether to omit empty `extra` objects. Initially set through the `ignoreEmptyExtra` option.

### `include`

Which log record properties to include in the output. Initially set through the `include` option.
