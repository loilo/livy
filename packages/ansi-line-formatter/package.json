{
  "name": "@livy/ansi-line-formatter",
  "type": "module",
  "version": "1.0.4",
  "description": "Formats Livy log records as single lines with terminal highlighting",
  "keywords": [
    "logging",
    "logger",
    "logs",
    "log",
    "loglevel",
    "tools",
    "monolog",
    "formatter",
    "ansi"
  ],
  "author": "loilo <florian.reuschel@gmx.de>",
  "homepage": "https://github.com/loilo/livy/tree/master/packages/ansi-line-formatter#readme",
  "license": "MIT",
  "exports": {
    "types": "./lib/ansi-line-formatter.d.ts",
    "default": "./lib/ansi-line-formatter.js"
  },
  "sideEffects": false,
  "directories": {
    "lib": "lib",
    "test": "test"
  },
  "files": [
    "lib"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/loilo/livy/tree/master/packages/ansi-line-formatter",
  "scripts": {
    "build": "del lib && tsc --declaration",
    "lint:readme": "remark --quiet --no-stdout -u validate-links README.md",
    "lint:source": "eslint \"src/**/*.ts\"",
    "test": "vitest run --coverage"
  },
  "bugs": {
    "url": "https://github.com/loilo/livy/issues"
  },
  "dependencies": {
    "@livy/contracts": "^1.1.2",
    "@livy/line-formatter": "^1.0.4",
    "@livy/util": "^1.0.6",
    "@types/luxon": "^3.2.0",
    "@types/node": "^18.11.18",
    "chalk": "^5.2.0",
    "chalk-template": "^1.1.0",
    "emphasize": "^6.0.0",
    "luxon": "^3.2.0"
  },
  "devDependencies": {
    "typescript": "^5.1.0",
    "@livy/bin": "file:../../bin",
    "@livy/test-utils": "file:../../test-utils"
  },
  "gitHead": "bd96a3e4d1e5870873b0e0b9e538486aea6343a2"
}
