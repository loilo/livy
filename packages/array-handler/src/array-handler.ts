import type {
  LogRecord,
  ResettableInterface,
  SyncHandlerInterface,
} from '@livy/contracts'
import { AbstractLevelBubbleHandler } from '@livy/util/handlers/abstract-level-bubble-handler'
import { MirrorSyncHandlerMixin } from '@livy/util/handlers/mirror-sync-handler-mixin'
import { ProcessableHandlerMixin } from '@livy/util/handlers/processable-handler-mixin'

/**
 * Stores log records in an array; great for debugging
 */
export class ArrayHandler
  extends MirrorSyncHandlerMixin(
    ProcessableHandlerMixin(AbstractLevelBubbleHandler),
  )
  implements SyncHandlerInterface, ResettableInterface
{
  private _records: LogRecord[] = []

  /**
   * @inheritdoc
   */
  public handleSync(record: LogRecord) {
    this._records.push(record)
    return !this.bubble
  }

  /**
   * @inheritdoc
   */
  public handleBatchSync(records: LogRecord[]) {
    this._records.push(...records)
  }

  /**
   * Get the stored records
   */
  public get records() {
    return this._records
  }

  /**
   * @inheritdoc
   */
  public reset() {
    this._records.splice(0)
  }
}
