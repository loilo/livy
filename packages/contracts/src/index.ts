export * from './closable-handler-interface.js'
export * from './formattable-handler-interface.js'
export * from './formatter-interface.js'
export * from './handler-interface.js'
export * from './log-level.js'
export * from './log-record.js'
export * from './logger-interface.js'
export * from './processable-handler-interface.js'
export * from './processor-interface.js'
export * from './resettable-interface.js'
