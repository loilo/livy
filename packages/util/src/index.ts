export type * from './formatters/abstract-batch-formatter.js'
export type * from './formatters/included-record-properties.js'
export type * from './formatters/line-formatter.js'
export type * from './handlers/abstract-batch-handler.js'
export type * from './handlers/abstract-formatting-processing-handler.js'
export type * from './handlers/abstract-level-bubble-handler.js'
export type * from './handlers/formattable-handler-mixin.js'
export type * from './handlers/is-closable-handler-interface.js'
export type * from './handlers/is-formattable-handler-interface.js'
export type * from './handlers/is-processable-handler-interface.js'
export type * from './handlers/is-sync-handler-interface.js'
export type * from './handlers/mirror-sync-handler-mixin.js'
export type * from './handlers/processable-handler-mixin.js'
export type * from './handlers/respect-level-mixin.js'
export type * from './console-adapter-interface.js'
export type * from './console-adapter.js'
export type * from './environment.js'
export type * from './gated-set.js'
export type * from './helpers.js'
export type * from './html.js'
export type * from './is-resettable-interface.js'
export type * from './mixin.js'
export type * from './timer.js'
export type * from './types.js'
export type * from './validatable-set.js'
