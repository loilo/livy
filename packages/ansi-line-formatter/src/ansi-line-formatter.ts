import type { LogLevel } from '@livy/contracts'
import { LineFormatter, LineFormatterOptions } from '@livy/line-formatter'
import { isEmpty } from '@livy/util/helpers'
import chalk, { supportsColor } from 'chalk'
import { DateTime } from 'luxon'

import { emphasize } from 'emphasize'
import chalkTemplate from 'chalk-template'

export interface AnsiLineFormatterOptions extends LineFormatterOptions {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   * If not provided (or set to `undefined`), color support is going
   * to be automatically detected
   */
  decorated: boolean | undefined
}

/**
 * Formats log records as single lines with terminal highlighting
 */
export class AnsiLineFormatter extends LineFormatter {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   */
  public decorated: boolean | undefined

  public constructor({
    decorated,
    ...options
  }: Partial<AnsiLineFormatterOptions> = {}) {
    super(options)
    this.decorated = decorated
  }

  /**
   * @inheritdoc
   */
  protected formatDatetime(datetime: DateTime) {
    const timestamp = super.formatDatetime(datetime)

    if (this.shouldDecorate()) {
      return chalk.dim(timestamp)
    } else {
      return timestamp
    }
  }

  /**
   * @inheritdoc
   */
  protected formatLevel(level: LogLevel) {
    const formatted = super.formatLevel(level)

    if (!this.shouldDecorate()) {
      return formatted
    } else {
      let color: string
      switch (level) {
        case 'emergency':
          color = 'red'
          break

        case 'alert':
          color = 'red'
          break

        case 'critical':
          color = 'red'
          break

        case 'error':
          color = 'red'
          break

        case 'warning':
          color = 'yellow'
          break

        case 'notice':
          color = 'blue'
          break

        case 'info':
          color = 'blue'
          break

        case 'debug':
          color = 'blue.dim'
          break

        /* c8 ignore next 2: This should never happen, but is included for type safety */
        default:
          return formatted
      }

      return chalkTemplate`{${color} ${formatted}}`
    }
  }

  /**
   * @inheritdoc
   */
  protected formatData(object: any, ignoreEmpty: boolean) {
    if (isEmpty(object) && ignoreEmpty) {
      return ''
    }

    const stringified = super.formatData(object)

    if (this.shouldDecorate()) {
      return emphasize.highlight('json', stringified).value
    } else {
      return stringified
    }
  }

  /**
   * Check whether the formatter should use ANSI codes to decorate log records
   */
  public shouldDecorate() {
    return Boolean(
      this.decorated !== undefined
        ? this.decorated
        : supportsColor
        ? supportsColor.hasBasic
        : false,
    )
  }
}
