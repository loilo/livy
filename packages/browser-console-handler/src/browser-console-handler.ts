import type { LogLevel, LogRecord, SyncHandlerInterface } from '@livy/contracts'
import { SeverityMap } from '@livy/contracts'
import {
  AbstractLevelBubbleHandler,
  AbstractLevelBubbleHandlerOptions,
} from '@livy/util/handlers/abstract-level-bubble-handler'
import { MirrorSyncHandlerMixin } from '@livy/util/handlers/mirror-sync-handler-mixin'
import { ProcessableHandlerMixin } from '@livy/util/handlers/processable-handler-mixin'
import { isEmpty } from '@livy/util/helpers'
import type { DateTime } from 'luxon'

export interface BrowserConsoleHandlerOptions
  extends AbstractLevelBubbleHandlerOptions {
  /**
   * Whether to use the browser's built-in console.debug() for the "debug" level,
   * which is only visible in the dev tools when explicitly configured so
   */
  useNativeDebug: boolean

  /**
   * Whether to include timestamps in the output
   */
  timestamps: boolean

  /**
   * The Console object to use
   */
  console: Console
}

/**
 * Writes log records to a browser console
 */
export class BrowserConsoleHandler
  extends MirrorSyncHandlerMixin(
    ProcessableHandlerMixin(AbstractLevelBubbleHandler),
  )
  implements SyncHandlerInterface
{
  private useNativeDebug: boolean
  private timestamps: boolean
  private console: Console

  public constructor({
    useNativeDebug = false,
    timestamps = false,
    console = self.console,
    ...options
  }: Partial<BrowserConsoleHandlerOptions> = {}) {
    super(options)

    this.console = console
    this.timestamps = timestamps
    this.useNativeDebug = useNativeDebug
  }

  /**
   * @inheritdoc
   */
  protected formatDatetime(datetime: DateTime) {
    return ['color: #888', datetime.toFormat('HH:mm:ss.SSS')]
  }

  /**
   * @inheritdoc
   */
  protected formatLevel(level: LogLevel) {
    const formatted = level.toUpperCase()

    let color: string
    switch (level) {
      case 'emergency':
        color = 'background-color: #f43a63; padding: 0 0.4em; color: #fff'
        break

      case 'alert':
        color = 'background-color: #f43a63; padding: 0 0.4em; color: #fff'
        break

      case 'critical':
        color = 'color: #f43a63'
        break

      case 'error':
        color = 'color: #f43a63'
        break

      case 'warning':
        color = 'color: #ffaa2b'
        break

      case 'notice':
        color = 'color: rgb(60, 161, 224)'
        break

      case 'info':
        color = 'color: rgb(60, 161, 224)'
        break

      case 'debug':
      default:
        color = 'color: rgba(60, 161, 224, 0.7)'
        break
    }

    return [color, formatted]
  }

  /**
   * @inheritdoc
   */
  public handleSync(record: LogRecord) {
    if (!this.isHandling(record.level)) {
      return
    }

    const resetter = 'color: inherit; background-color: inherit'
    const parameters: any[] = [record.message]
    const hasExtra = !isEmpty(record.extra)

    if (hasExtra || !isEmpty(record.context)) {
      parameters.push(record.context)
    }

    if (hasExtra) {
      parameters.push(record.extra)
    }

    const formattedTimestamp = this.timestamps
      ? [...this.formatDatetime(record.datetime), resetter]
      : []
    const formattedLevel = [...this.formatLevel(record.level), resetter]

    let placeholders = ['%c%s']
    if (this.timestamps) {
      placeholders.push('%c%s')
    }

    placeholders = placeholders.map(placeholder => `${placeholder}%c`)

    if (record.severity <= SeverityMap.error) {
      this.console.error(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else if (record.severity === SeverityMap.warning) {
      this.console.warn(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else if (record.severity <= SeverityMap.info) {
      this.console.info(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else {
      if (this.useNativeDebug) {
        placeholders.pop()

        this.console.debug(
          placeholders.join(' '),
          ...formattedTimestamp,
          ...parameters,
        )
      } else {
        this.console.log(
          placeholders.join(' '),
          ...formattedTimestamp,
          ...formattedLevel,
          ...parameters,
        )
      }
    }

    return !this.bubble
  }
}
