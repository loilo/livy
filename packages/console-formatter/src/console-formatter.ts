import { AnsiLineFormatter } from '@livy/ansi-line-formatter'
import { EOL } from '@livy/util/environment'
import { isEmpty } from '@livy/util/helpers'
import chalk from 'chalk'
import { dump } from 'js-yaml'
import { emphasize } from 'emphasize'
import chalkTemplate from 'chalk-template'

/**
 * Formats log records with highlighting for terminals in a human-readable way
 */
export class ConsoleFormatter extends AnsiLineFormatter {
  /**
   * Format a record's context object
   *
   * @param context The context to format
   */
  protected formatContext(context: any, ignoreEmpty: boolean) {
    const formattedContext = this.formatData(context, ignoreEmpty)
    if (formattedContext.length === 0) {
      return ''
    }

    return chalkTemplate`${EOL}  {dim Context:}${EOL}${formattedContext}`
  }

  /**
   * Format a record's extra object
   *
   * @param context The extra to format
   */
  protected formatExtra(extra: any, ignoreEmpty: boolean) {
    const formattedExtra = this.formatData(extra, ignoreEmpty)
    if (formattedExtra.length === 0) {
      return ''
    }

    return chalkTemplate`${EOL}  {dim Extra:}${EOL}${formattedExtra}`
  }

  /**
   * Clone a value but replace any occurrences of `value` in it
   *
   * @param container   The container to strip `value` from
   * @param value       The value to replace
   * @param replacement The string to put in place of undefined
   * @param cache       The internal cache used to avoid cycles
   */
  private cloneAndReplace(
    container: any,
    value: any,
    replacement: string,
    cache = new Map(),
  ): any {
    if (container === value) {
      return replacement
    }

    if (cache.has(container)) {
      return cache.get(container)
    }

    if (Array.isArray(container)) {
      const result = container.map(entry =>
        this.cloneAndReplace(entry, value, replacement, cache),
      )
      cache.set(container, result)
      return result
    }

    if (typeof container === 'object' && container !== null) {
      const copy = { ...container }
      for (const key in copy) {
        copy[key] = this.cloneAndReplace(copy[key], value, replacement, cache)
      }
      cache.set(container, copy)

      return copy
    }

    return container
  }

  /**
   * @inheritdoc
   */
  protected formatData(object: any, ignoreEmpty: boolean) {
    if (isEmpty(object)) {
      /* c8 ignore next: This is not used throughout the library, but can be useful when extending this formatter */
      return super.formatData(object, ignoreEmpty)
    } else {
      // We need to replace `undefined` and `null` values because they are
      // useful being represented but they cannot be serialized as YAML.
      // Since we merely use YAML as a human-readable data visualization tool here,
      // it's completely fine to have output that would not be properly parseable.

      // Generate a random alphanumeric string as replacement
      // for occurrences of undefined/null
      // Prefix with a letter ("x") to avoid random strings starting with a number
      // from being interpreted as a number by the YAML highlighter (which
      // prevents proper replacement in the end)
      const generateRandomString = () =>
        'x' +
        Math.random().toString(36).slice(2, 15) +
        Math.random().toString(36).slice(2, 15)

      const undefinedReplacement = generateRandomString()
      const nullReplacement = generateRandomString()

      // Replace all undefined/null occurrences in the object
      // with the generated random strings
      let sanitizedObject = object
      sanitizedObject = this.cloneAndReplace(
        sanitizedObject,
        undefined,
        undefinedReplacement,
      )
      sanitizedObject = this.cloneAndReplace(
        sanitizedObject,
        null,
        nullReplacement,
      )

      let serializedYaml = dump(sanitizedObject, {
        skipInvalid: true,
        indent: 2,
        flowLevel: 2,
      })

      // Empty object string may result from skipped invalid data like JS functions
      if (serializedYaml.trim() === '{}') {
        return super.formatData({}, ignoreEmpty)
      }

      if (this.shouldDecorate()) {
        serializedYaml = emphasize.highlight('yaml', serializedYaml).value
      }

      // Replace the generated random strings with a dimmed "undefined"/"null"
      serializedYaml = serializedYaml
        .replaceAll(
          new RegExp(undefinedReplacement, 'g'),
          chalk.gray('undefined'),
        )
        .replaceAll(new RegExp(nullReplacement, 'g'), chalk.gray('null'))

      return this.indent(serializedYaml)
    }
  }

  /**
   * Indents a (multiline) string with the given padding
   *
   * @param text The string to be indented
   * @param pad  The padding to be prepended on each line
   * @return string
   */
  protected indent(text: string, pad = '  ') {
    return text.replaceAll(/^/gm, pad)
  }
}
