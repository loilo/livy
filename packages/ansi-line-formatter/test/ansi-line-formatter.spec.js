import { describe, beforeEach, afterEach, it, expect, vi } from 'vitest'
import chalk, { supportsColor } from 'chalk'

vi.mock(
  'chalk',
  livyTestGlobals.getMockedModule(
    import('@livy/test-utils/mocks/modules/chalk.js'),
  ),
)

import { AnsiLineFormatter } from '../src/ansi-line-formatter'

const { record } = livyTestGlobals

describe('@livy/ansi-line-formatter', () => {
  beforeEach(() => {
    supportsColor.hasBasic = true
  })

  afterEach(() => {
    chalk.__reset()
  })

  it('should correctly derive color support from chalk', () => {
    supportsColor.hasBasic = true
    expect(new AnsiLineFormatter().shouldDecorate()).toBe(true)
    supportsColor.hasBasic = false
    expect(new AnsiLineFormatter().shouldDecorate()).toBe(false)
  })

  it('should respect the "decorated" option', () => {
    supportsColor.hasBasic = true
    expect(new AnsiLineFormatter({ decorated: false }).shouldDecorate()).toBe(
      false,
    )
    supportsColor.hasBasic = false
    expect(new AnsiLineFormatter({ decorated: true }).shouldDecorate()).toBe(
      true,
    )
  })

  it('should format records correctly without decoration', () => {
    const ansiLineFormatter = new AnsiLineFormatter({
      decorated: false,
      include: { channel: true },
    })

    const createRecord = record.with({
      message: 'Test AnsiLineFormatter',
      context: { foo: 1 },
      extra: { bar: 2 },
    })

    expect(
      ansiLineFormatter.format(createRecord({ level: 'debug' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'info' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'notice' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'warning' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'error' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'critical' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'alert' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'emergency' })),
    ).toMatchSnapshot()
  })

  it('should decorate records correctly', () => {
    const ansiLineFormatter = new AnsiLineFormatter({
      decorated: true,
      include: { channel: true },
    })

    const createRecord = record.with({
      message: 'Test LineFormatter',
      context: { foo: 1 },
      extra: { bar: 2 },
    })

    expect(
      ansiLineFormatter.format(createRecord({ level: 'debug' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'info' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'notice' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'warning' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'error' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'critical' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'alert' })),
    ).toMatchSnapshot()
    expect(
      ansiLineFormatter.format(createRecord({ level: 'emergency' })),
    ).toMatchSnapshot()
  })
})
