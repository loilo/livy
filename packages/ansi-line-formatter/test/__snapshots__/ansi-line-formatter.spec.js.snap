// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@livy/ansi-line-formatter > should decorate records correctly 1`] = `"[2m2015-08-15 14:30:00 [22mlogs [34m[2mDEBUG [22m[39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should decorate records correctly 2`] = `"[2m2015-08-15 14:30:00 [22mlogs [34mINFO [39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should decorate records correctly 3`] = `"[2m2015-08-15 14:30:00 [22mlogs [34mNOTICE [39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should decorate records correctly 4`] = `"[2m2015-08-15 14:30:00 [22mlogs [33mWARNING [39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should decorate records correctly 5`] = `"[2m2015-08-15 14:30:00 [22mlogs [31mERROR [39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should decorate records correctly 6`] = `"[2m2015-08-15 14:30:00 [22mlogs [31mCRITICAL [39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should decorate records correctly 7`] = `"[2m2015-08-15 14:30:00 [22mlogs [31mALERT [39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should decorate records correctly 8`] = `"[2m2015-08-15 14:30:00 [22mlogs [31mEMERGENCY [39mTest LineFormatter {[33m\\"foo\\"[39m:[36m1[39m} {[33m\\"bar\\"[39m:[36m2[39m}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 1`] = `"2015-08-15 14:30:00 logs DEBUG Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 2`] = `"2015-08-15 14:30:00 logs INFO Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 3`] = `"2015-08-15 14:30:00 logs NOTICE Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 4`] = `"2015-08-15 14:30:00 logs WARNING Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 5`] = `"2015-08-15 14:30:00 logs ERROR Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 6`] = `"2015-08-15 14:30:00 logs CRITICAL Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 7`] = `"2015-08-15 14:30:00 logs ALERT Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;

exports[`@livy/ansi-line-formatter > should format records correctly without decoration 8`] = `"2015-08-15 14:30:00 logs EMERGENCY Test AnsiLineFormatter {\\"foo\\":1} {\\"bar\\":2}"`;
